components/brave_new_tab_ui/helpers/colorUtil.ts

```diff
@@ -0,0 +1,173 @@
+  getRelativeLuminance = () => {
+    if (!this.isValid()) {
+        console.error(`Invalid color value: ${this}`)
+        return 0.0
+    }
+
+    let r = this.r
+    let g = this.g
+    let b = this.b
+    if (r === undefined || g === undefined || b === undefined) {
+        console.error(`Couldn't extract RGB of: ${this}`)
+        return 0.0
+    }
+
+    const cachedLuminance = this.cachedLuminance
+    if (cachedLuminance) {
+      return cachedLuminance
+    }
+
+    // Step 1. Lerp each channel to a value between 0 and 1.
+    [r, g, b] = [r, g, b].map(v => v / 255);
+
+    // Step 2. Adjust each channel's value.
+    [r, g, b] = [r, g, b].map(v => {
+        if (v <= 0.03928) {
+            v /= 12.92
+        } else {
+            v = ((v + 0.055) / 1.055) ** 2.4
+        }
+        return v
+    })
+
+    // Step 3. Get Luminance.
+    const cacheKey = this.luminanceCacheKey
+    ColorData.#luminanceCache.set(cacheKey, 0.2126 * r + 0.7152 * g + 0.0722 * b)
+    return ColorData.#luminanceCache.get(cacheKey)!
+  }
+}
+
+/**
+ * Returns contrast ratio of two colors based on a definition from
+// https://www.w3.org/TR/WCAG20/#contrast-ratiodef
+ * @param {ColorData} a - ColorData
+ * @param {ColorData} b - ColorData
+ * @returns Contrast ratio of two colors
+ */
+export function getContrastRatio (a: ColorData, b: ColorData): number {
+  const [lighter, darker] = a.getRelativeLuminance() > b.getRelativeLuminance() ? [a, b] : [b, a]
+  return (lighter.getRelativeLuminance() + 0.05) / (darker.getRelativeLuminance() + 0.05)
+}
+
+// This value is much lower than that of WCAG 2.0. https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0#qr-visual-audio-contrast-contrast
+// The value was decided by our UX team.
+let thresholdForReadability = 1.5
+
+export function setThresholdForReadability (threshold: number) {
+  thresholdForReadability = threshold
+}
+
+export function getThresholdForReadability () {
+  return thresholdForReadability
+}
+
+// Determines if . The threshold we use
+/**
+ * It returns true if |foreground| is readable on |background|.
+ * @param {ColorData} background - The background color
+ * @param {ColorData} foreground - The foreground color.
+ * @returns {boolean} - True if the contrast ratio is greater than or equal to the threshold for readability
+ */
+export function isReadable (background: ColorData, foreground: ColorData) {
+  return getContrastRatio(background, foreground) >= thresholdForReadability
+}
+
+/**
+ * Convenient function to get readability for our background type
+ * @param {NewTab.BackgroundWallpaper | undefined} background -
+ * NewTab.BackgroundWallpaper | undefined
+ * @returns {boolean} - true if background is readable
+ */
+export default function isReadableOnBackground (background: NewTab.BackgroundWallpaper | undefined) {
+  if (!background || background.type !== 'color') {
+    // Consider other cases readable. We can't determine the color from images.
```

Comment:
It'd be cool if we could do something like get the average darkness or something. Probably way outside the scope of this PR though :laughing: